### DISCUSSION TITLE
Using pyFAST in Google CoLab and GitHub Actions

### DISCUSSION THREAD

**Discussion by andreped** [FAST-Imaging/FAST#185] [Q&A]
I have previously been able to use pyFAST in Docker containers, meaning that pyFAST can be used on headless servers.
Deployment was then performed in pyFAST using `fast.Pipeline`.

Recently, FAST added headless support to `runPipeline` which would mean that I could replace the temporary custom pyFAST `fast.Pipeline` approach I had with a more generalized CLI method.

I therefore wanted to test using this CLI in two relevant use cases:

1. In a GitHub Actions workflow for unit testing (CPU only - run in containers hosted by GitHub)
2. In a CoLab Jupyter Notebook to produce interactive demos

For the first one, as there is no GPU available, I used PoCL, which at least made it possible for me to import pyFAST without any errors. However, when running some simple tests, I got some errors, which I did not get on my local setup - likely due to the usage of PoCL. I therefore put my focus on CoLab instead.

For the latter, I tried to use PoCL as well (also choosing CPU runtime), just to see if I could get a CPU-only approach working.  I ran some simple `runPipeline` and python tutorial tests. None of these worked. One of them executed without errors, but it did not seem like it actually started running.

I have made a [gist](https://colab.research.google.com/gist/andreped/afcd631afa37839bbb129ed61d734777/fast-python-pocl.ipynb) to reproduce these experiments, which could be valuable to test PoCL further with pyFAST on CoLab.

Of course, in CoLab, we do actually have access to a GPU, so PoCL may not be required, however, setting up OpenCL/OpenGL in CoLab seemed tricky. I could make a proper attempt though, but I was not tempted.

I guess the main reason is that pyFAST is not really compatible with PoCL, which may not be that surprising.

Anyways, I'm making this thread for discussion as I'm debugging this further.

**Comment by andreped** [comment]
I managed to get pyFAST working on CoLab using the same install setup as I used for setting up `pyFAST` in Docker (see [here](https://colab.research.google.com/gist/andreped/b6b0850a6e9349efbbd2501087cc663d/fast-python-opengl.ipynb)).
So that is great! That means that we can make pyFAST demos in CoLab which is a great way for others to learn about FAST :]
Note that the nuclei segmentation pipeline still didn't start properly. No idea why. Any ideas?

I made the [gist](https://colab.research.google.com/gist/andreped/b6b0850a6e9349efbbd2501087cc663d/fast-python-opengl.ipynb) openly available for others to try.

But of course, we do not have access to a GPU when running pyFASTs on CPU clusters such as using GitHub-hosted containers through GitHub Actions. Any alternative to PoCL that I could try?

**Comment by smistad** [comment]
The errors you get with pocl are build errors. Meaning that pocl fails to compile some of FAST's OpenCL code during execution. This is most likely due to bugs in pocl.

You can try newer versions of pocl (build yourself or use newer Ubuntu) or you can try Intel's CPU only OpenCL runtime

**Reply by andreped** [reply]
I managed to find some precompiled binaries of PoCL [here](https://repology.org/project/pocl/versions), as suggested from the PoCL README

However, it seems like only the latest and very specific Ubuntu versions support newer PoCL versions. Sadly, CoLab runs with Ubuntu 22.04 (jammy), which then uses PoCL 1.8, and there is no easy way to change that.
<img width="1173" alt="Screenshot 2023-09-03 at 20 02 04" src="https://github.com/smistad/FAST/assets/29090665/9d347741-7228-4008-a69a-5b945526f4f4">

I guess the best solution would be to build the binaries ourselves for the relevant operating system(s) and version(s) and release them, but ugh... Annoying...

**Reply by andreped** [reply]
Regarding `Intel's CPU only OpenCL Runtime`, I have never gotten it working, hence why I turned to PoCL which at least seemed to work during installation. From `clinfo` it tells me that no OpenCL platforms were found.

Made a [gist](https://colab.research.google.com/gist/andreped/9c1d2da73165a19faf1a58c331add2ba/fast-python-intel-opencl-runtime.ipynb) that reproduces the issue. Maybe there is some linking problem? Any ideas?

Or hmm, do you need a dedicated Intel GPU to use this Runtime (see [here](https://github.com/intel/compute-runtime/issues/628#issuecomment-1470085717))? Right now, I set the runtime in CoLab to CPU only. Doubt there is an Intel GPU option anyways. Wouldn't be the case for GitHub Actions clusters either, I would think.

**Reply by andreped** [reply]
As I realized that the wrong OpenCL CPU RT was downloaded, I have tried updating the gist but I still run into the same issue.

See updated [gist](https://colab.research.google.com/gist/andreped/c4ddfb132345f1c57b888fe58e326db9/fast-python-intel-opencl-runtime-fix.ipynb).

**Reply by andreped** [reply]
I managed to fix the issue. After decompression, I needed to place the extracted `x64/` directory at the appropriate location. Now `clinfo` finds the platform.

However, I now run into another issue when running tests (for both `runPipeline` and using `pyFAST`):
```
INFO [133801215184576] Found 1 OpenCL platforms.
INFO [133801215184576] 1 platforms selected for inspection.
INFO [133801215184576] Platform 0: Intel(R) Corporation
INFO [133801215184576] This platform has 1 available devices in total
INFO [133801215184576] Looking for GPU devices only.
INFO [133801215184576] 0 selected.
INFO [133801215184576] NO suitable GPU found! Trying to find other devices
INFO [133801215184576] Found 1 OpenCL platforms.
INFO [133801215184576] 1 platforms selected for inspection.
INFO [133801215184576] Platform 0: Intel(R) Corporation
INFO [133801215184576] This platform has 1 available devices in total
INFO [133801215184576] Looking for all types of devices.
INFO [133801215184576] 1 selected.
INFO [133801215184576] Inspecting device 0 with the name            Intel(R) Xeon(R) CPU @ 2.20GHz
INFO [133801215184576] The device was accepted.
INFO [133801215184576]            Intel(R) Xeon(R) CPU @ 2.20GHz has 2 compute units
INFO [133801215184576] The device            Intel(R) Xeon(R) CPU @ 2.20GHz got a score of 2
INFO [133801215184576] The platform Intel(R) OpenCL was selected as the best platform.
INFO [133801215184576] A total of 1 devices were selected for the context from this platform:
INFO [133801215184576] The best device was:            Intel(R) Xeon(R) CPU @ 2.20GHz
ERROR [133801215184576] Terminated with unhandled exception: clCreateContext
```

See updated [gist](https://colab.research.google.com/gist/andreped/21749b8f89983b97622719ba4f9c9bbf/fast-python-intel-opencl-runtime-fix.ipynb).

**Reply by smistad** [reply]
Strange. Now it is able to find the platform and the CPU, but fails when initializing OpenCL..

**Reply by andreped** [reply]
Strange indeed. But I added the updated gist, so you can debug that further on CoLab, when you have the time.

When this is properly working, I will try to run some tests on GitHub-hosted servers. If that works, then I believe it opens some new opportunities :]

**Reply by andreped** [reply]
As it seemed like I was potentially using the wrong Intel OpenCL package, I now tried installing the **Intel OpenCL oneAPI runtime** instead to see if I am able to use `FAST` on a CPU-only server.

After installation, it seems to be able to find the platforms, but we get `RuntimeError: clCreateImage` when attempting to run any `FAST` pipelines with `runPipeline` or `pyFAST` scripts.

To reproduce the error, see [gist](https://colab.research.google.com/gist/andreped/0031872ee6b18d764da52c98585f3fd6/fast-python-intel-oneapi-runtime-opencl.ipynb) on CoLab.

**Reply by lomshabhishek** [reply]
On Mac M1 as well I faced similar issue

```
Traceback (most recent call last):
  File "/Users/abhisheklomsh/Downloads/pyfast_file.py", line 46, in <module>
    .run()
  File "/Users/abhisheklomsh/.pyenv/versions/3.10.13/lib/python3.10/site-packages/fast/fast.py", line 3440, in run
    return _fast.ProcessObject_run(self, executeToken)
RuntimeError: Thread stopped because: clCreateImage
```
